rust/theoros/src/services/hyperlane/mod.rs

```diff
@@ -30,59 +32,153 @@ impl Service for HyperlaneService {
         }
     }
 
+    /// Processes validator checkpoints by fetching signed checkpoints from all validators for each unsigned nonce.
+    ///
+    /// This function performs the following steps:
+    ///
+    /// 1. **Retrieve Unsigned Nonces**:
+    ///    - Fetches all the nonces currently stored in the `UnsignedCheckpointsStorage`.
+    ///
+    /// 2. **Retrieve Validators and Fetchers**:
+    ///    - Gets all registered validators and their corresponding fetchers from the `ValidatorsFetchersStorage`.
+    ///
+    /// 3. **Fetch Signed Checkpoints**:
+    ///    - Attempts to fetch the signed checkpoint all unsigned nonce from each validator's fetcher (in parallel),
+    ///
+    /// 4. **Process Completed Nonces**:
+    ///    - After all fetches are completed, iterates over the unsigned nonces again.
+    ///    - Checks if all validators have signed the nonce using the `all_validators_signed_nonce` method.
+    ///    - **Note**: Currently, the function only proceeds if **all** validators have signed the nonce.
+    ///      - There's a `TODO` to modify this behavior to use a quorum method (e.g., consider a nonce as valid if 66% of validators have signed it).
+    ///    - If all validators have signed the nonce:
+    ///        - Calls `store_event_updates(nonce)` to process and store the updates associated with that nonce.
+    ///        - Removes the nonce from the `UnsignedCheckpointsStorage`, as it has been fully processed.
+    ///
     async fn process_validator_checkpoints(&self) {
-        let validators_locations = self.state.storage.validators_locations().all().await;
-        let futures = validators_locations
-            .into_iter()
-            .map(|(validator, fetcher)| self.process_single_validator(validator, fetcher));
+        let unsigned_nonces = self.storage.unsigned_checkpoints().nonces().await;
+        if unsigned_nonces.is_empty() {
+            return;
+        }
 
+        let validators_fetchers = self.storage.validators_fetchers().all().await;
+        let mut futures = Vec::with_capacity(unsigned_nonces.len());
+        for &nonce in &unsigned_nonces {
+            for (validator, fetcher) in &validators_fetchers {
+                let fut = self.fetch_checkpoint_for_validator(*validator, fetcher.clone(), nonce);
+                futures.push(fut);
+            }
+        }
         futures::future::join_all(futures).await;
+
+        // NOTE: At the moment, we only process updates when ALL validators have signed a message.
+        // TODO: We should instead use a quorum method - if 66% have signed, consider it ok.
+        let validator_addresses: Vec<Felt> = validators_fetchers.keys().cloned().collect();
+        for &nonce in &unsigned_nonces {
+            if !self.all_validators_signed_nonce(&validator_addresses, nonce).await {
+                continue;
+            }
+            // TODO: If the nonce n+1 is fully signed, shall we ignore every nonces before..? Or raise an alert?
```

Comment:
In the current behavior we loop through all unsigned nonces this can lead to some perf bottleneck.

As you mention in this comment, we should probably process in descending order (right now it's ascending) and then remove all previous unsigned checkpoints.

Reply:
So this would be true in theory if the signed were checkpoints almost instantly. What you describe only happens at the startup of Theoros (when we index from `current_block - 5`).

It takes about 10~15 second for a checkpoint to be signed so what will happens 100% of the time is:
* Unsigned checkpoints [1, 2, 3],
* _We check every 1 seconds_
* After ~10 seconds, checkpoint 1 is fully signed and stored
* Unsigned checkpoints [2, 3],
* etc...

So in this setup, iterating from 3 would just be a lost of time, unless there's really strange and the update don't occur every seconds, it should never happen.

So yeah I think iterating backwards would be actually reducing perfs

Reply:
makes sense, let's do it then